### PR TITLE
Feature/idp

### PR DESCRIPTION
**Pull request by lkwdwrd**
This is the initial implementation to add identity features to the SDK. This builds on the previous refactor managing the client and config for Storage and Identity separately.

In implementing, several bugs in notes were discovered and update. Previously premium note features would not work -- in fact any note options would not work, and notes were only written with the same writer and recipient. This addresses those bugs.

Still to do:
- Proper documentation, internal and in the readme
- Unit testing for new features

### CODE REVIEW

File: lib/identity/config.js [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+   */
+  static fromObject(obj) {
+    if (typeof obj === 'string') {
+      try {
+        obj = JSON.parse(obj)
+      } catch (err) {
+        throw new Error('Config.fromObject param JSON string could not be parsed.')
+      }
+    }
+    const realmId = obj.realmId || obj.api_key_id
```
Comment (galxy25) [inline lib/identity/config.js]: `|| obj.realm_id`

File: lib/identity/config.js [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    const realmName = obj.realmName || obj.api_secret
```
Comment (galxy25) [inline lib/identity/config.js]: `|| obj.realm_name`

File: lib/identity/index.js [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+import CryptoConsumer from '../utils/cryptoConsumer'
+import { Config as StorageConfig } from '../storage'
+import { PublicKey, SigningKey } from '../types'
+
+async function deriveNoteCreds(id, username, password) {
+  const noteID = await id.crypto.genericHash(username + id.config.realmId)
+  const cryptoKeyPair = await id.crypto.deriveCryptoKey(
+    password,
+    id.config.realmId + id.config.realmName,
+    10000
```
Comment (galxy25) [inline lib/identity/index.js]: extract into constant

File: lib/identity/index.js [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+  )
+  const signingKeyPair = await id.crypto.deriveSigningKey(
+    password,
+    cryptoKeyPair.publicKey + cryptoKeyPair.privateKey,
+    10000
```
Comment (galxy25) [inline lib/identity/index.js]: extract into constant

File: lib/identity/index.js [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+   * @returns {Client} The storage Client constructor.
+   */
+  get StorageClient() {
+    // Use this.constructor to ensure we referencing the implementing class, not an interface class.
+    return this.constructor.StorageClient
+  }
+
+  async register(username, password, token) {
+    const cryptoKeys = await this.StorageClient.generateKeypair()
+    const signingKeys = await this.StorageClient.generateSigningKeypair()
```
Comment (galxy25) [inline lib/identity/index.js]: nit: these key generation methods should live on the top level Tozny/helper object versus the StorageClient.

File: lib/identity/index.js [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+    /* eslint-enable */
+    const request = await fetch(this.config.apiUrl + '/v1/identity/register', {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/json'
+      },
+      body: JSON.stringify(payload)
+    })
+    const response = await checkStatus(request)
+    const json = await response.json()
```
Comment (galxy25) [inline lib/identity/index.js]: really want to see check status and parse JSON steps consolidated into a helper method that will allow us to dry out the repeated steps of constructing the request, making it, checking the status and parsing the response json.

File: lib/identity/index.js (hunk 5 above)
Comment (lkwdwrd) [reply lib/identity/index.js]: They need access to the crypto object. We could put together a smarter helper object. `helpers` was intended mostly to be platform specific helpers, such as a way of pulling a profile from the file system ~/.tozny folder. Makes sens in node, but not at all in the browser context. At the very lease these should stay in the storage Client for backwards compatibility reasons. I'm thinking we should more directly expose the crypto implementation on the top level Tozny object? Will figure out something :)

File: lib/identity/index.js (hunk 6 above)
Comment (lkwdwrd) [reply lib/identity/index.js]: good point, this pattern is used frequently. I'll explore putting these together.

File: lib/identity/index.js (hunk 5 above)
Comment (lkwdwrd) [reply lib/identity/index.js]: This interface does not directly determine the shape of an implementation's export. I'm wondering if it somehow should. Perhaps a default method in index.js which takes all the pieces and exports the 'Tozny' object. This helps ensure the default package exports keep the same shape. Still not sure the best shape for that, yet, though. I'm going to leave it looser with an eye towards tightening it as we get to a stable release state.

In the mean time, I have exposed the crypto implementation in the browser SDK directly from the Tozny object:

```
keyPair = await Tozny.crypto.generateKeyPair() / o/r any other defined crypto operation available there
```

File: lib/identity/index.js (hunk 6 above)
Comment (lkwdwrd) [reply lib/identity/index.js]: added the helper and use it in this case -- other cases can adopt as they are modified.

**Pull request merged**